### PR TITLE
Fix serialization P2 issues

### LINKED ISSUE #16639: Serialization: Cannot load into empty/blank target
**Describe the bug**
When I try to dump and load data in metabase `v1.39.3`, it throws an exception in `Inserting User` step. So I can't restore data.

**Logs**
```
2021-06-16 10:59:37,443 INFO serialization.cmd :: BEGIN LOAD from dump with context {:mode :update, :on-error :abort}
2021-06-16 10:59:37,464 INFO serialization.upsert :: Inserting User
2021-06-16 10:59:37,485 ERROR serialization.cmd :: ERROR LOAD from dump: Assert failed: ((every-pred string? (complement str/blank?)) password)
java.lang.AssertionError: Assert failed: ((every-pred string? (complement str/blank?)) password)
        at metabase.models.user$pre_insert.invokeStatic(user.clj:28)
        at metabase.models.user$pre_insert.invoke(user.clj:25)
        at toucan.models$fn__24435$fn__24469$G__24442__24474.invoke(models.clj:178)
        at toucan.models$do_pre_insert.invokeStatic(models.clj:329)
        at toucan.models$do_pre_insert.invoke(models.clj:324)
        at toucan.db$insert_BANG_.invokeStatic(db.clj:610)
        at toucan.db$insert_BANG_.invoke(db.clj:597)
        at metabase_enterprise.serialization.upsert$insert_many_individually_BANG_$iter__57434__57438$fn__57439$fn__57440.invoke(upsert.clj:82)
        at metabase_enterprise.serialization.upsert$insert_many_individually_BANG_$iter__57434__57438$fn__57439.invoke(upsert.clj:80)
        at clojure.lang.LazySeq.sval(LazySeq.java:42)
        at clojure.lang.LazySeq.seq(LazySeq.java:51)
        at clojure.lang.RT.seq(RT.java:535)
        at clojure.core$seq__5402.invokeStatic(core.clj:137)
        at clojure.core$map$fn__5870.invoke(core.clj:2757)
        at clojure.lang.LazySeq.sval(LazySeq.java:42)
        at clojure.lang.LazySeq.seq(LazySeq.java:58)
        at clojure.lang.RT.seq(RT.java:535)
        at clojure.core$seq__5402.invokeStatic(core.clj:137)
        at clojure.core$concat$cat__5495$fn__5496.invoke(core.clj:734)
        at clojure.lang.LazySeq.sval(LazySeq.java:42)
        at clojure.lang.LazySeq.seq(LazySeq.java:51)
        at clojure.lang.RT.seq(RT.java:535)
        at clojure.core$seq__5402.invokeStatic(core.clj:137)
        at clojure.core$sort.invokeStatic(core.clj:3113)
        at clojure.core$sort_by.invokeStatic(core.clj:3119)
        at clojure.core$sort_by.invokeStatic(core.clj:3119)
        at clojure.core$sort_by.invoke(core.clj:3119)
        at metabase_enterprise.serialization.upsert$maybe_upsert_many_BANG_.invokeStatic(upsert.clj:144)
        at metabase_enterprise.serialization.upsert$maybe_upsert_many_BANG_.invoke(upsert.clj:119)
        at metabase_enterprise.serialization.load$fn__58123.invokeStatic(load.clj:641)
        at metabase_enterprise.serialization.load$fn__58123.invoke(load.clj:638)
        at clojure.lang.MultiFn.invoke(MultiFn.java:234)
        at metabase_enterprise.serialization.cmd$fn__58635$load__58640$fn__58641.invoke(cmd.clj:52)
        at metabase_enterprise.serialization.cmd$fn__58635$load__58640.invoke(cmd.clj:39)
        at clojure.lang.Var.invoke(Var.java:388)
        at metabase.cmd$load.invokeStatic(cmd.clj:150)
        at metabase.cmd$load.doInvoke(cmd.clj:141)
        at clojure.lang.RestFn.applyTo(RestFn.java:139)
        at clojure.core$apply.invokeStatic(core.clj:665)
        at clojure.core$apply.invoke(core.clj:660)
        at metabase.cmd$run_cmd$fn__82234.invoke(cmd.clj:191)
        at metabase.cmd$run_cmd.invokeStatic(cmd.clj:191)
        at metabase.cmd$run_cmd.invoke(cmd.clj:187)
        at clojure.lang.Var.invoke(Var.java:388)
        at metabase.core$run_cmd.invokeStatic(core.clj:149)
        at metabase.core$run_cmd.invoke(core.clj:147)
        at metabase.core$_main.invokeStatic(core.clj:171)
        at metabase.core$_main.doInvoke(core.clj:166)
        at clojure.lang.RestFn.applyTo(RestFn.java:137)
        at metabase.core.main(Unknown Source)
```

**To Reproduce**
1. Run metabase by `java -jar metabase-v1.39.3.jar`.
2. Create a user according to Web UI.
3. Stop metabase and dump data by `java -jar metabase-v1.39.3.jar dump ./dump --user [my mail address]`
4. Remove existing h2 database by `rm -rf metabase.db.*`
5. Load dump data by `java -jar metabase-v1.39.3.jar load ./dump --mode update --on-error abort`
6. See the error above

**Expected behavior**
Should restore all dump data.

**Information about your Metabase Installation:**

```
{
  "browser-info": {
    "language": "en-US",
    "platform": "MacIntel",
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.101 Safari/537.36",
    "vendor": "Google Inc."
  },
  "system-info": {
    "file.encoding": "UTF-8",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "11.0.9.1+1",
    "java.vendor": "AdoptOpenJDK",
    "java.vendor.url": "https://adoptopenjdk.net/",
    "java.version": "11.0.9.1",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "11.0.9.1+1",
    "os.name": "Mac OS X",
    "os.version": "10.16",
    "user.language": "en",
    "user.timezone": "America/Los_Angeles"
  },
  "metabase-info": {
    "databases": [
      "h2"
    ],
    "hosting-env": "unknown",
    "application-database": "h2",
    "application-database-details": {
      "database": {
        "name": "H2",
        "version": "1.4.197 (2018-03-18)"
      },
      "jdbc-driver": {
        "name": "H2 JDBC Driver",
        "version": "1.4.197 (2018-03-18)"
      }
    },
    "run-mode": "prod",
    "version": {
      "date": "2021-05-27",
      "tag": "v1.39.3",
      "branch": "release-x.39.x",
      "hash": "598a112"
    },
    "settings": {
      "report-timezone": null
    }
  }
}
```

**Severity**
Can workaround by using a previous version.

**Additional context**
It doesn't happen in `v1.39.2`

### LINKED ISSUE #16931: Serialization crashes on dump if there's no collections
**Describe the bug**
When trying to create a dump without using "--user" flags, the process exits with an error

**Logs**
With PostgreSQL:
<details>
2021-07-07 21:28:08,395 INFO serialization.cmd :: BEGIN DUMP to /target via user null
org.postgresql.util.PSQLException: ERROR: syntax error at or near ")"
  Position: 36
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2553)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2285)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:323)
	at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:473)
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:393)
	at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:164)
	at org.postgresql.jdbc.PgPreparedStatement.executeQuery(PgPreparedStatement.java:114)
	at com.mchange.v2.c3p0.impl.NewProxyPreparedStatement.executeQuery(NewProxyPreparedStatement.java:431)
	at clojure.java.jdbc$execute_query_with_params.invokeStatic(jdbc.clj:1090)
	at clojure.java.jdbc$execute_query_with_params.invoke(jdbc.clj:1084)
	at clojure.java.jdbc$db_query_with_resultset_STAR_.invokeStatic(jdbc.clj:1113)
	at clojure.java.jdbc$db_query_with_resultset_STAR_.invoke(jdbc.clj:1093)
	at clojure.java.jdbc$query.invokeStatic(jdbc.clj:1182)
	at clojure.java.jdbc$query.invoke(jdbc.clj:1144)
	at toucan.db$query.invokeStatic(db.clj:288)
	at toucan.db$query.doInvoke(db.clj:284)
	at clojure.lang.RestFn.invoke(RestFn.java:410)
	at toucan.db$simple_select.invokeStatic(db.clj:394)
	at toucan.db$simple_select.invoke(db.clj:383)
	at toucan.db$select.invokeStatic(db.clj:662)
	at toucan.db$select.doInvoke(db.clj:656)
	at clojure.lang.RestFn.invoke(RestFn.java:423)
	at metabase_enterprise.serialization.cmd$select_collections.invokeStatic(cmd.clj:107)
	at metabase_enterprise.serialization.cmd$select_collections.invoke(cmd.clj:92)
	at metabase_enterprise.serialization.cmd$dump.invokeStatic(cmd.clj:144)
	at metabase_enterprise.serialization.cmd$dump.invoke(cmd.clj:116)
	at metabase_enterprise.serialization.cmd$dump.invokeStatic(cmd.clj:119)
	at metabase_enterprise.serialization.cmd$dump.invoke(cmd.clj:116)
	at clojure.lang.Var.invoke(Var.java:388)
	at metabase.cmd$dump.invokeStatic(cmd.clj:161)
	at metabase.cmd$dump.doInvoke(cmd.clj:154)
	at clojure.lang.RestFn.invoke(RestFn.java:423)
	at metabase.cmd$dump.invokeStatic(cmd.clj:157)
	at metabase.cmd$dump.invoke(cmd.clj:154)
	at clojure.lang.AFn.applyToHelper(AFn.java:154)
	at clojure.lang.RestFn.applyTo(RestFn.java:132)
	at clojure.core$apply.invokeStatic(core.clj:665)
	at clojure.core$apply.invoke(core.clj:660)
	at metabase.cmd$run_cmd$fn__82234.invoke(cmd.clj:191)
	at metabase.cmd$run_cmd.invokeStatic(cmd.clj:191)
	at metabase.cmd$run_cmd.invoke(cmd.clj:187)
	at clojure.lang.Var.invoke(Var.java:388)
	at metabase.core$run_cmd.invokeStatic(core.clj:149)
	at metabase.core$run_cmd.invoke(core.clj:147)
	at metabase.core$_main.invokeStatic(core.clj:171)
	at metabase.core$_main.doInvoke(core.clj:166)
	at clojure.lang.RestFn.applyTo(RestFn.java:137)
	at metabase.core.main(Unknown Source)
Command failed with exception: ERROR: syntax error at or near ")"
  Position: 36
</details>

With MySQL
<details>
2021-07-07 22:53:45,766 INFO serialization.cmd :: BEGIN DUMP to /target via user null
java.sql.SQLSyntaxErrorException: (conn=131) You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near ') AND `archived` = FALSE)' at line 1
	at org.mariadb.jdbc.internal.util.exceptions.ExceptionFactory.createException(ExceptionFactory.java:62)
	at org.mariadb.jdbc.internal.util.exceptions.ExceptionFactory.create(ExceptionFactory.java:153)
	at org.mariadb.jdbc.MariaDbStatement.executeExceptionEpilogue(MariaDbStatement.java:274)
	at org.mariadb.jdbc.ClientSidePreparedStatement.executeInternal(ClientSidePreparedStatement.java:229)
	at org.mariadb.jdbc.ClientSidePreparedStatement.execute(ClientSidePreparedStatement.java:149)
	at org.mariadb.jdbc.ClientSidePreparedStatement.executeQuery(ClientSidePreparedStatement.java:163)
	at com.mchange.v2.c3p0.impl.NewProxyPreparedStatement.executeQuery(NewProxyPreparedStatement.java:431)
	at clojure.java.jdbc$execute_query_with_params.invokeStatic(jdbc.clj:1090)
	at clojure.java.jdbc$execute_query_with_params.invoke(jdbc.clj:1084)
	at clojure.java.jdbc$db_query_with_resultset_STAR_.invokeStatic(jdbc.clj:1113)
	at clojure.java.jdbc$db_query_with_resultset_STAR_.invoke(jdbc.clj:1093)
	at clojure.java.jdbc$query.invokeStatic(jdbc.clj:1182)
	at clojure.java.jdbc$query.invoke(jdbc.clj:1144)
	at toucan.db$query.invokeStatic(db.clj:288)
	at toucan.db$query.doInvoke(db.clj:284)
	at clojure.lang.RestFn.invoke(RestFn.java:410)
	at toucan.db$simple_select.invokeStatic(db.clj:394)
	at toucan.db$simple_select.invoke(db.clj:383)
	at toucan.db$select.invokeStatic(db.clj:662)
	at toucan.db$select.doInvoke(db.clj:656)
	at clojure.lang.RestFn.invoke(RestFn.java:423)
	at metabase_enterprise.serialization.cmd$select_collections.invokeStatic(cmd.clj:107)
	at metabase_enterprise.serialization.cmd$select_collections.invoke(cmd.clj:92)
	at metabase_enterprise.serialization.cmd$dump.invokeStatic(cmd.clj:144)
	at metabase_enterprise.serialization.cmd$dump.invoke(cmd.clj:116)
	at metabase_enterprise.serialization.cmd$dump.invokeStatic(cmd.clj:119)
	at metabase_enterprise.serialization.cmd$dump.invoke(cmd.clj:116)
	at clojure.lang.Var.invoke(Var.java:388)
	at metabase.cmd$dump.invokeStatic(cmd.clj:161)
	at metabase.cmd$dump.doInvoke(cmd.clj:154)
	at clojure.lang.RestFn.invoke(RestFn.java:423)
	at metabase.cmd$dump.invokeStatic(cmd.clj:157)
	at metabase.cmd$dump.invoke(cmd.clj:154)
	at clojure.lang.AFn.applyToHelper(AFn.java:154)
	at clojure.lang.RestFn.applyTo(RestFn.java:132)
	at clojure.core$apply.invokeStatic(core.clj:665)
	at clojure.core$apply.invoke(core.clj:660)
	at metabase.cmd$run_cmd$fn__82234.invoke(cmd.clj:191)
	at metabase.cmd$run_cmd.invokeStatic(cmd.clj:191)
	at metabase.cmd$run_cmd.invoke(cmd.clj:187)
	at clojure.lang.Var.invoke(Var.java:388)
	at metabase.core$run_cmd.invokeStatic(core.clj:149)
	at metabase.core$run_cmd.invoke(core.clj:147)
	at metabase.core$_main.invokeStatic(core.clj:171)
	at metabase.core$_main.doInvoke(core.clj:166)
	at clojure.lang.RestFn.applyTo(RestFn.java:137)
	at metabase.core.main(Unknown Source)
Caused by: org.mariadb.jdbc.internal.util.exceptions.MariaDbSqlException: You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near ') AND `archived` = FALSE)' at line 1
	at org.mariadb.jdbc.internal.util.exceptions.MariaDbSqlException.of(MariaDbSqlException.java:34)
	at org.mariadb.jdbc.internal.protocol.AbstractQueryProtocol.exceptionWithQuery(AbstractQueryProtocol.java:194)
	at org.mariadb.jdbc.internal.protocol.AbstractQueryProtocol.exceptionWithQuery(AbstractQueryProtocol.java:177)
	at org.mariadb.jdbc.internal.protocol.AbstractQueryProtocol.executeQuery(AbstractQueryProtocol.java:321)
	at org.mariadb.jdbc.ClientSidePreparedStatement.executeInternal(ClientSidePreparedStatement.java:220)
	... 43 more
Caused by: java.sql.SQLException: You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near ') AND `archived` = FALSE)' at line 1
	at org.mariadb.jdbc.internal.protocol.AbstractQueryProtocol.readErrorPacket(AbstractQueryProtocol.java:1688)
	at org.mariadb.jdbc.internal.protocol.AbstractQueryProtocol.readPacket(AbstractQueryProtocol.java:1550)
	at org.mariadb.jdbc.internal.protocol.AbstractQueryProtocol.getResult(AbstractQueryProtocol.java:1513)
	at org.mariadb.jdbc.internal.protocol.AbstractQueryProtocol.executeQuery(AbstractQueryProtocol.java:318)
	... 44 more
Command failed with exception: (conn=131) You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near ') AND `archived` = FALSE)' at line 1
</details>

**To Reproduce**
1) Create a Metabase instance
2) try to do a dump without any user flags
3) see the error

**Expected behavior**
The dump works or we give our users a warning, but we don't fail with a SQL error

**Screenshots**
Logs attached

**Information about your Metabase Installation:**
- Your browser and the version: Brave latest
- Your operating system: Ubuntu LTS latest
- Your databases: None
- Metabase version: .39.3
- Metabase hosting environment: Docker
- Metabase internal database: tried with both MySQL and Postgres

**Severity**
It breaks serialization if not using a --user, so users that are not using this flag have their serialization process broken

**Additional context**
Reported by a person on a trial
EDIT: just saw the DB logs and it says
```
postgres_origin         | 2021-07-08 13:44:02.751 UTC [50] ERROR:  syntax error at or near ")" at character 36
postgres_origin         | 2021-07-08 13:44:02.751 UTC [50] STATEMENT:  SELECT * FROM "collection" WHERE (() AND "archived" = FALSE)
```
Also, when running this script that we have in our documentation the process fails even with --user flag
```console
create_users()
{
    ADMIN_EMAIL=${MB_ADMIN_EMAIL:-admin@metabase.local}
    ADMIN_PASSWORD=${MB_ADMIN_PASSWORD:-Metapass123}

    METABASE_HOST=${1}
    METABASE_PORT=${2}

    echo "⌚︎ Waiting for Metabase to start"
    while (! curl -s -m 5 http://${METABASE_HOST}:${METABASE_PORT}/api/session/properties -o /dev/null); do sleep 5; done

    echo "😎 Creating admin user"

    SETUP_TOKEN=$(curl -s -m 5 -X GET \
        -H "Content-Type: application/json" \
        http://${METABASE_HOST}:${METABASE_PORT}/api/session/properties \
        | jq -r '.["setup-token"]'
    )

    MB_TOKEN=$(curl -s -X POST \
        -H "Content-type: application/json" \
        http://${METABASE_HOST}:${METABASE_PORT}/api/setup \
        -d '{
        "token": "'${SETUP_TOKEN}'",
        "user": {
            "email": "'${ADMIN_EMAIL}'",
            "first_name": "Metabase",
            "last_name": "Admin",
            "password": "'${ADMIN_PASSWORD}'"
        },
        "prefs": {
            "allow_tracking": false,
            "site_name": "Metawhat"
        }
    }' | jq -r '.id')


    echo -e "\n👥 Creating some basic users: "
    curl -s "http://${METABASE_HOST}:${METABASE_PORT}/api/user" \
        -H 'Content-Type: application/json' \
        -H "X-Metabase-Session: ${MB_TOKEN}" \
        -d '{"first_name":"Basic","last_name":"User","email":"basic@somewhere.com","login_attributes":{"region_filter":"WA"},"password":"'${ADMIN_PASSWORD}'"}'

    curl -s "http://${METABASE_HOST}:${METABASE_PORT}/api/user" \
        -H 'Content-Type: application/json' \
        -H "X-Metabase-Session: ${MB_TOKEN}" \
        -d '{"first_name":"Basic 2","last_name":"User","email":"basic2@somewhere.com","login_attributes":{"region_filter":"CA"},"password":"'${ADMIN_PASSWORD}'"}'

    echo -e "\n👥 Basic users created!"
}

# remove any existing containers
docker rm --force metabase-origin postgres-origin metabase-target postgres-target
docker volume prune --force

# export METABASE_VERSION=v1.36.7         # version from example - works ok
export METABASE_VERSION=v1.39.3       # version in use - errors on dump
# export METABASE_VERSION=v1.39.4       # latest version - errors on dump

echo "Testing serialisation with metabase-enterprise:${METABASE_VERSION}"
echo "======================================================"

docker network create metanet

docker run --rm -d --name postgres-origin \
    -p 5433:5432 \
    -e POSTGRES_USER=metabase \
    -e POSTGRES_PASSWORD=knockknock \
    --network metanet \
    postgres:13.3-alpine

docker run --rm -d --name metabase-origin \
    -p 5001:3000 \
    -e MB_DB_TYPE=postgres \
    -e MB_DB_DBNAME=metabase \
    -e MB_DB_PORT=5432 \
    -e MB_DB_USER=metabase \
    -e MB_DB_PASS=knockknock \
    -e MB_DB_HOST=postgres-origin \
    --network metanet \
    metabase/metabase-enterprise:${METABASE_VERSION}

create_users localhost 5001

# target instance

docker run --rm -d --name postgres-target \
    -p 5434:5432 \
    -e POSTGRES_USER=metabase \
    -e POSTGRES_PASSWORD=knockknock \
    --network metanet \
    postgres:13.3-alpine

docker run --rm -d --name metabase-target \
    -p 5002:3000 \
    -e MB_DB_TYPE=postgres \
    -e MB_DB_DBNAME=metabase \
    -e MB_DB_PORT=5432 \
    -e MB_DB_USER=metabase \
    -e MB_DB_PASS=knockknock \
    -e MB_DB_HOST=postgres-target \
    --network metanet \
    metabase/metabase-enterprise:${METABASE_VERSION}

create_users localhost 5002

# try an export
METABASE_DUMP_DIR=metabase-serialisation-test

rm -r ${METABASE_DUMP_DIR}
mkdir ${METABASE_DUMP_DIR}          # Metabase creates it with root ownership if we don't create it ourselves
chmod 777 ${METABASE_DUMP_DIR}

docker run --rm --name metabase-dump --network metanet \
    -e MB_DB_CONNECTION_URI="postgres://postgres-origin:5432/metabase?user=metabase&password=knockknock" \
    -e MUID=$(id -u) \
    -e MGID=$(id -g) \
    -v "$PWD/${METABASE_DUMP_DIR}:/target" metabase/metabase-enterprise:${METABASE_VERSION} "dump /target --user admin@metabase.local"
```

but when doing the process raising the following docker-compose file and then doing the steps manually it works with the --user flag:
```
version: '3.7'
services:
  metabase_origin:
    image: metabase/metabase-enterprise:v1.39.3
    container_name: metabase_origin
    volumes: 
      - $PWD/metabase_dump:/target
    ports:
      - 3001:3000
    environment: 
      - "MB_DB_DBNAME=metabase"
      - "MB_DB_USER=metabase"
      - "MB_DB_PASS=mysecretpassword"
      - "MB_DB_PORT=5432"
      - "MB_DB_HOST=postgres"
      - "MB_DB_TYPE=postgres"
    networks: 
      - metanet1
    depends_on: 
      - postgres_origin
  postgres_origin:
    image: postgres:13.3-alpine
    container_name: postgres_origin
    hostname: postgres
    environment:
      - "POSTGRES_USER=metabase"
      - "POSTGRES_DB=metabase"
      - "POSTGRES_PASSWORD=mysecretpassword"
    volumes:
      - $PWD/postgres_origin:/var/lib/postgresql/data
    networks: 
      - metanet1
  metabase_destination:
    image: metabase/metabase-enterprise:v1.39.3
    container_name: metabase_destination
    volumes: 
      - $PWD/metabase_dump:/target
    ports:
      - 3002:3000
    environment: 
      - "MB_DB_DBNAME=metabase"
      - "MB_DB_USER=metabase"
      - "MB_DB_PASS=mysecretpassword"
      - "MB_DB_PORT=5432"
      - "MB_DB_TYPE=postgres"
      - "MB_DB_HOST=postgres"
    networks: 
      - metanet2
    depends_on: 
      - postgres_destination
  postgres_destination:
    image: postgres:13.3-alpine
    container_name: postgres_destination
    hostname: postgres
    environment:
      - "POSTGRES_USER=metabase"
      - "POSTGRES_DB=metabase"
      - "POSTGRES_PASSWORD=mysecretpassword"
    volumes:
      - $PWD/postgres_destination:/var/lib/postgresql/data
    networks: 
      - metanet2
networks: 
  metanet1:
    driver: bridge
  metanet2:
    driver: bridge
```

### LINKED ISSUE #16954: Serialization `--mode skip` incorrectly updates some objects
**Describe the bug**
When loading a source into a target, then `--mode skip` (default) runs as if it was `--mode update` on some objects.

Regression since 1.38.1, which used another run on cards to fix some issues. This is might not needed anymore, since there's the "second pass" since 1.39.3, which fixed a lot more issues.

**To Reproduce**
1. On source, run `dump` process
2. On target, run `load` (or `load --mode skip`) process
3. On source, example modify the query of a question, or change the cards on a dashboard - redo step 1+2
4. When running the load process the second time, then it will also note that it's updating cards (or inserting dashcards)

<details><summary>Second load process</summary>

```
2021-07-09 16:24:27,919 INFO serialization.cmd :: BEGIN LOAD from source with context {:mode :skip, :on-error :continue}
2021-07-09 16:24:27,929 INFO serialization.load :: Processing collection at path source/collections/root
2021-07-09 16:24:27,963 INFO serialization.upsert :: Collection "mode-skip test - collection" (ID 28) already exists -- skipping
2021-07-09 16:24:27,965 INFO serialization.load :: Processing collection at path source/collections/root/collections/mode-skip test - collection
2021-07-09 16:24:28,028 INFO serialization.upsert :: Card "mode-skip test - SQL" (ID 62) already exists -- skipping
2021-07-09 16:24:28,075 INFO serialization.upsert :: Updating Card "mode-skip test - SQL" (ID 62)
2021-07-09 16:24:28,107 INFO driver.impl :: Registered driver :h2 (parents: [:sql-jdbc]) 🚚
Load driver :h2 took 22.7 ms
2021-07-09 16:24:28,123 INFO driver.impl :: Initializing driver :sql...
2021-07-09 16:24:28,125 INFO driver.impl :: Initializing driver :sql-jdbc...
2021-07-09 16:24:28,126 INFO driver.impl :: Initializing driver :h2...
2021-07-09 16:24:28,170 INFO serialization.upsert :: Dashboard "mode-skip test - dashboard" (ID 22) already exists -- skipping
2021-07-09 16:24:28,204 INFO serialization.upsert :: Inserting DashboardCard
2021-07-09 16:24:28,205 INFO serialization.upsert :: Inserting DashboardCard
2021-07-09 16:24:28,206 INFO serialization.upsert :: DashboardCard 72 already exists -- skipping
2021-07-09 16:24:28,224 INFO serialization.cmd :: END LOAD from source with context {:mode :skip, :on-error :continue}
```

</details>

**Expected behavior**
Skip should never touch existing objects if mode=skip.

**Information about your Metabase Installation:**
Tested 1.38.0 thru 1.40.0. Regression since 1.38.1

**Severity**
The most general Serialization use-case is to update, so giving it P2, but it's close to P1.

### PR DESCRIPTION
**Pull request by jeff303**
# Serialization P2 Fixes August 2021
## Commit 1: Fix serialization dump error when there are no collections
- Update `select-collections` to correctly handle the case where there are no collections
- Adding new test that removes all collections, then ensures that dump works with no errors

## Commit 2: Fix serialization load error into empty/blank target DB
- Loads dump into completely blank target app DB
- If user is newly created by the load process, they are assigned a random password and a password reset email is generated and sent to them

## Commit 3: Fix serialization objects being incorrectly updated on skip
- Gets rid of unneeded `maybe-fixup-card-template-ids!` fn, which is what was forcing update of some entities even on `:skip`

### CODE REVIEW

**Review by kidd — APPROVED** [review]
lgtm :+1:

File: src/metabase/models/metric.clj [hunk 1]
```diff
@@ -17,6 +17,9 @@
 
 (models/defmodel Metric :metric)
 
+(defn- pre-delete [{:keys [id]}]
+  (db/delete! 'Dependency :model "Metric", :model_id id))
```
[the remark this answers is not in the document]
Comment (flamber) [reply src/metabase/models/metric.clj:21]: Segments are similar to Metrics, but none can be deleted - only retired/archived. Perhaps I'm not understanding what this code is supposed to do.

**Review by flamber — CHANGES_REQUESTED** [review]
Testing `ebc9dcd`

When I do a load without any mode-parameter (or mode=skip), then a lot of things stops working. And it still seems to be updating a dashboard, when I re-run the load without mode (#16954)

Without mode=update, it seems like it's breaking 14914 and 14909 and EE533 and EE532 and EE487 - and possibly other nested things?
Also seems to convert some GUI nested questions to SQL - haven't completely understood what's happening here.
Doesn't update the SQL correctly (`-- DUMMY QUERY FOR SERIALIZATION FIRST PASS INSERT`)

When using mode=update, then it seems like it's not correctly removing the existing cards on the dashboard, so all updated cards gets duplicated (I updated the text with "UPDATED" to the two first textboxes, and changed a question to reference another question, and then added a new card)
image

I'm just going to block this for now, since it has an approve.

File: src/metabase/models/metric.clj (hunk 1 above)
[the remark this answers is not in the document]
Comment (jeff303) [reply src/metabase/models/metric.clj:21]: There likely are other places.  However, the "hard delete" path only seems to be available via tests, which is what led to the error this is intended to fix.  The actual production workflow doesn't do a hard delete, just an archive.  So this change will only impact tests.

**Comment by jeff303** [comment]
I continue to be troubled at how difficult it is to "see" the same behavior from the backend tests, which are clearly belie the state of the functionality.  Once I narrow down on what is causing this latest round of discrepancies, it might be time to revamp the test suite yet again.

**Comment by jeff303** [comment]
OK, here is the fundamental issue at play, I believe.  It all comes down to the `identity-condition` as seen [here](https://github.com/metabase/metabase/blob/master/enterprise/backend/src/metabase_enterprise/serialization/upsert.clj#L32).  This is the set of fields that will be checked to determine if an entity loaded from the YAML is the same as an existing one in the target app DB.

If the load process reads an entity `A` from YAML, and does not find a "matching" target entity, then `A` will be inserted as a new record into the target DB.  This is true regardless of `--mode`, because the process considers it to be "new".  So in the case of `DashboardCard`, for instance, changing the text in a card will make the load process consider the YAML entity to be new.  That's because the text is technically under the `visualization_settings`, [which **is** part](https://github.com/metabase/metabase/blob/master/enterprise/backend/src/metabase_enterprise/serialization/upsert.clj#L40) of the `identity-condition`.

The `--mode` flag only controls what happens on an identity match (not a miss, which - as outlined above - will always result in an insert). If set to `:update` then the entity will be updated if any other of its fields (i.e. things not in the `identity-condition`) changed.  If not, then it will also be skipped.

One idea would be to simply remove `visualization_settings` from the `identity-condition` for `DashboardCard`, but I don't think that would work.  These text cards only _have_ difference there (the `:card-id` is `nil`).

What we could do is always drop all existing DashboardCard instances `DC` for a Dashboard `D`, if and only if mode is `:update` and we are about to update `D`.  Then all loaded YAML `DC` instances will simply replace everything that was there.  But we will need to think about whether this makes sense, and whether the idea should be extended to other entities.

**Review by flamber — APPROVED** [review]
Tested `3a3e5fb` - it works great :tada: - now :taco: :man_dancing: :wink: after a little lint :broom: 

We probably want to create an issue about https://github.com/metabase/metabase/pull/17388#issuecomment-900415001, so we can figure out how to handle dashcards better (removal, Textbox, multiple of the same card_id, ...)
Do you wanna do it @jeff303, since you got the best technical context?

**Comment by rlotun** [comment]
@metabase-bot backport release-x.40.x

**Pull request merged**